File: melbourne/graphics/scoreboard.py
import logging
from multiprocessing.pool import ThreadPool
from pathlib import Path

from pathvalidate import sanitize_filename
from skia import Surface

from melbourne.graphics.config import (
    ScoreboardColors,
    ScoreboardConfig,
    ScoreboardFonts,
    ScoreboardSizes,
    load_fonts,
)
from melbourne.graphics.utilities import (
    TextAlignment,
    draw_image,
    draw_line,
    draw_rectangle,
    draw_text,
    load_image,
)

logger = logging.getLogger(__name__)


def _generate_scoreboard(
    config: ScoreboardConfig,
    fonts: ScoreboardFonts,
    colors: ScoreboardColors,
    voter_idx: int,
):
    """Generate a single scoreboard"""
    sizes = ScoreboardSizes(fonts, config, voter_idx)
    surface = Surface.MakeRasterN32Premul(sizes.width, sizes.height)
    canvas = surface.getCanvas()

    # Draw scoreboard background
    canvas.clear(colors.background)

    # Draw voter details
    draw_rectangle(
        canvas,
        (0, 0),
        sizes.width,
        30.0 * sizes.scaling_ratio,
        colors.voter_header,
    )
    draw_text(
        canvas,
        point=(10.0 * sizes.scaling_ratio, 15.0 * sizes.scaling_ratio),
        text=f"Now Voting: {config.contest.voter_names[voter_idx]} ({voter_idx + 1}/{config.contest.num_voters})",
        font=fonts.voter_header,
        color=colors.voter_header_text,
        alignment=TextAlignment.LEFT,
    )

    # Draw contest title
    draw_rectangle(
        canvas,
        (0, 30 * sizes.scaling_ratio),
        sizes.width,
        30.0 * sizes.scaling_ratio,
        colors.contest_header,
    )
    draw_text(
        canvas,
        point=(10.0 * sizes.scaling_ratio, 45.0 * sizes.scaling_ratio),
        text=(
            f"{config.title} Results"
            if config.append_results_to_title
            else config.title
        ),
        font=fonts.contest_header,
        color=colors.contest_header_text,
        alignment=TextAlignment.LEFT,
    )

    # Draw background rectangles for entry details
    num_left_col_entries = (
        int(config.contest.num_entries / 2) + config.contest.num_entries % 2
    )
    num_right_col_entries = config.contest.num_entries - num_left_col_entries
    draw_rectangle(
        canvas,
        point=(10 * sizes.scaling_ratio, 70 * sizes.scaling_ratio),
        width=sizes.rectangle,
        height=35 * sizes.scaling_ratio * num_left_col_entries,
        fill_color=colors.entry_details,
        stroke_width=2.0,
        stroke_color=colors.entry_details_border,
    )
    draw_rectangle(
        canvas,
        point=(
            20.0 * sizes.scaling_ratio + sizes.rectangle,
            70.0 * sizes.scaling_ratio,
        ),
        width=sizes.rectangle,
        height=35.0 * sizes.scaling_ratio * num_right_col_entries,
        fill_color=colors.entry_details,
        stroke_width=2.0,
        stroke_color=colors.entry_details_border,
    )

    # Draw entry details
    entries = config.contest.get_results_after_voter(voter_idx)
    for i, entry in enumerate(entries):
        if i < num_left_col_entries:
            x_offset = 0.0
            y_offset = i
        else:
            x_offset = 10.0 * sizes.scaling_ratio + sizes.rectangle
            y_offset = i - num_left_col_entries

        # Draw the entry's country flag
        if config.display_flags:
            if config.custom_flags_dir and entry.flag.lower().startswith("custom/"):
                flag = entry.flag.split("/", maxsplit=1)[-1]
                flag_path = config.custom_flags_dir / flag
            else:
                flag_path = config.flags_dir / Path(entry.flag)

            if not flag_path.exists():
                logger.warning(f"Could not find flag at {flag_path=}")
            else:
                # Load the flag and scale width to 20.0 * scaling
                flag = load_image(flag_path, width=20.0 * sizes.scaling_ratio)

                if config.display_flag_borders:
                    x = 27.0 * sizes.scaling_ratio - flag.width() / 2.0 + x_offset
                    y = (
                        87.0 * sizes.scaling_ratio
                        - flag.height() / 2.0
                        + 35.0 * sizes.scaling_ratio * y_offset
                    )
                    draw_image(
                        canvas,
                        flag,
                        x,
                        y,
                        stroke_width=1,
                        stroke_color=colors.flag_border,
                    )

        # Draw entry details
        draw_text(
            canvas,
            point=(
                20.0 * sizes.scaling_ratio + x_offset + sizes.flag_offset,
                80.0 * sizes.scaling_ratio + 35.0 * sizes.scaling_ratio * y_offset,
            ),
            text=entry.country,
            font=fonts.country,
            color=colors.country_text,
            alignment=TextAlignment.LEFT,
        )
        draw_text(
            canvas,
            point=(
                20.0 * sizes.scaling_ratio + x_offset + sizes.flag_offset,
                94.0 * sizes.scaling_ratio + 35.0 * sizes.scaling_ratio * y_offset,
            ),
            text=f"{entry.artist} – {entry.song}",
            font=fonts.entry_details,
            color=colors.entry_details_text,
            alignment=TextAlignment.LEFT,
        )

        # Draw the entry's total number of received points
        total_pts_color = (
            colors.dqed_pts if entry.dq_statuses[voter_idx] else colors.total_pts
        )
        total_pts_text_color = (
            colors.dqed_pts_text
            if entry.dq_statuses[voter_idx]
            else colors.total_pts_text
        )
        draw_rectangle(
            canvas,
            point=(
                30.0 * sizes.scaling_ratio
                + x_offset
                + sizes.flag_offset
                + sizes.entry_details,
                77.0 * sizes.scaling_ratio + 35.0 * sizes.scaling_ratio * y_offset,
            ),
            width=29.0 * sizes.scaling_ratio,
            height=20.0 * sizes.scaling_ratio,
            fill_color=total_pts_color,
        )
        draw_text(
            canvas,
            point=(
                44.5 * sizes.scaling_ratio
                + x_offset
                + sizes.flag_offset
                + sizes.entry_details,
                87.0 * sizes.scaling_ratio + 35.0 * sizes.scaling_ratio * y_offset,
            ),
            text=str(entry.display_pts[voter_idx]),
            font=fonts.total_pts,
            color=total_pts_text_color,
            alignment=TextAlignment.CENTER,
        )

        # Draw the entry's points received by the current voter
        if len(entry.votes[voter_idx]) > 0:
            draw_rectangle(
                canvas,
                point=(
                    59.0 * sizes.scaling_ratio
                    + x_offset
                    + sizes.flag_offset
                    + sizes.entry_details,
                    77.0 * sizes.scaling_ratio + 35.0 * sizes.scaling_ratio * y_offset,
                ),
                width=24.0 * sizes.scaling_ratio,
                height=20.0 * sizes.scaling_ratio,
                fill_color=colors.received_pts,
            )
            if entry.votes[voter_idx].endswith(".0"):
                received_votes_text = str(int(float(entry.votes[voter_idx])))
            else:
                received_votes_text = entry.votes[voter_idx]
            draw_text(
                canvas,
                point=(
                    71.0 * sizes.scaling_ratio
                    + x_offset
                    + sizes.flag_offset
                    + sizes.entry_details,
                    87.0 * sizes.scaling_ratio + 35.0 * sizes.scaling_ratio * y_offset,
                ),
                text=received_votes_text,
                font=fonts.received_pts,
                color=colors.received_pts_text,
                alignment=TextAlignment.CENTER,
            )

        # Draw a dividing line between entries
        draw_line(
            canvas,
            starting_point=(
                10.0 * sizes.scaling_ratio + x_offset,
                104.5 * sizes.scaling_ratio + 35.0 * sizes.scaling_ratio * y_offset,
            ),
            ending_point=(
                10.0 * sizes.scaling_ratio + x_offset + sizes.rectangle,
                104.5 * sizes.scaling_ratio + 35.0 * sizes.scaling_ratio * y_offset,
            ),
            color=colors.divider_line,
            stroke_width=0.5 * sizes.scaling_ratio,
        )

    # Save image to file
    image = surface.makeImageSnapshot()
    image_file_path = config.output_dir / sanitize_filename(
        f"{voter_idx + 1} - {config.contest.voter_names[voter_idx]}.png"
    )
    image.save(str(image_file_path))


def generate_scoreboards(config: ScoreboardConfig):
    """Generate scoreboards for the contest"""
    fonts = load_fonts(config.fonts_config)
    colors = ScoreboardColors(config.main_color, config.accent_color)

    inputs = [(config, fonts, colors, i) for i in range(config.contest.num_voters)]
    with ThreadPool(processes=5) as pool:
        pool.starmap(_generate_scoreboard, inputs)

File: Melbourne/Scoreboard/Scoreboard.cs
namespace Melbourne.Scoreboard;

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;
using SkiaSharp;

public static class Scoreboard
{
    
        private static string SanitizeFileName(string fileName)
        {
            foreach (var c in Path.GetInvalidFileNameChars())
                fileName = fileName.Replace(c, '_');
            return fileName;
        }

        public static string GenerateScoreboard(
            Config config,
            Fonts fonts,
            Colors colors,
            int voterIdx)
        {
            var sizes = new Sizes(fonts, config, voterIdx);
            float scaling = sizes.ScalingRatio;

            using var bitmap = new SKBitmap((int)sizes.Width, (int)sizes.Height);
            using var canvas = new SKCanvas(bitmap);

            // Clear background
            canvas.Clear(colors.Background);

            // Voter header
            Utilities.DrawRectangle(canvas, (0, 0), sizes.Width, 30f * scaling, fillColor: colors.VoterHeader);
            Utilities.DrawText(
                canvas,
                (10f * scaling, 15f * scaling),
                $"Now Voting: {config.Contest.VoterNames[voterIdx]} ({voterIdx + 1}/{config.Contest.NumVoters})",
                fonts.VoterHeader,
                colors.VoterHeaderText,
                Utilities.TextAlignment.Left);

            // Contest header
            Utilities.DrawRectangle(canvas, (0, 30f * scaling), sizes.Width, 30f * scaling, fillColor: colors.ContestHeader);
            Utilities.DrawText(
                canvas,
                (10f * scaling, 45f * scaling),
                config.AppendResultsToTitle ? $"{config.Title} Results" : config.Title,
                fonts.ContestHeader,
                colors.ContestHeaderText,
                Utilities.TextAlignment.Left);

            // Background rectangles for entries
            int numLeft = config.Contest.NumEntries / 2 + config.Contest.NumEntries % 2;
            int numRight = config.Contest.NumEntries - numLeft;

            Utilities.DrawRectangle(
                canvas,
                (10f * scaling, 70f * scaling),
                sizes.Rectangle,
                35f * scaling * numLeft,
                fillColor: colors.EntryDetails,
                strokeColor: colors.EntryDetailsBorder,
                strokeWidth: 2f);

            Utilities.DrawRectangle(
                canvas,
                (20f * scaling + sizes.Rectangle, 70f * scaling),
                sizes.Rectangle,
                35f * scaling * numRight,
                fillColor: colors.EntryDetails,
                strokeColor: colors.EntryDetailsBorder,
                strokeWidth: 2f);

            // Draw entry details
            var entries = config.Contest.ResultsAfterVoter(voterIdx);
            for (int i = 0; i < entries.Count; i++)
            {
                var entry = entries[i];
                float xOffset = i < numLeft ? 0f : 10f * scaling + sizes.Rectangle;
                float yOffset = i < numLeft ? i : i - numLeft;

                // Draw flag
                if (config.DisplayFlags)
                {
                    string flagPath = Path.Combine(config.FlagsDir, entry.Flag);
                    if (File.Exists(flagPath))
                    {
                        using var flag = Utilities.LoadImage(flagPath, 20f * scaling);
                        if (config.DisplayFlagBorders)
                        {
                            float fx = 27f * scaling - flag.Width / 2f + xOffset;
                            float fy = 87f * scaling - flag.Height / 2f + 35f * scaling * yOffset;
                            Utilities.DrawImage(canvas, flag, fx, fy, strokeColor: colors.FlagBorder, strokeWidth: 1f);
                        }
                        else
                        {
                            float fx = 27f * scaling - flag.Width / 2f + xOffset;
                            float fy = 87f * scaling - flag.Height / 2f + 35f * scaling * yOffset;
                            Utilities.DrawImage(canvas, flag, fx, fy);
                        }
                    }
                }

                // Entry text
                float baseX = 20f * scaling + xOffset + sizes.FlagOffset;
                Utilities.DrawText(canvas, (baseX, 80f * scaling + 35f * scaling * yOffset), entry.Country, fonts.Country, colors.CountryText, Utilities.TextAlignment.Left);
                Utilities.DrawText(canvas, (baseX, 94f * scaling + 35f * scaling * yOffset), $"{entry.Artist} – {entry.Song}", fonts.EntryDetails, colors.EntryDetailsText, Utilities.TextAlignment.Left);

                // Total points
                var totalPtsColor = entry.DqStatuses[voterIdx] ? colors.DqedPoints : colors.TotalPoints;
                var totalPtsTextColor = entry.DqStatuses[voterIdx] ? colors.DqedPointsText : colors.TotalPointsText;

                Utilities.DrawRectangle(
                    canvas,
                    (30f * scaling + xOffset + sizes.FlagOffset + sizes.EntryDetailsWidth,
                     77f * scaling + 35f * scaling * yOffset),
                    29f * scaling,
                    20f * scaling,
                    fillColor: totalPtsColor);

                Utilities.DrawText(
                    canvas,
                    (44.5f * scaling + xOffset + sizes.FlagOffset + sizes.EntryDetailsWidth,
                     87f * scaling + 35f * scaling * yOffset),
                    entry.DisplayPoints[voterIdx].ToString(),
                    fonts.TotalPoints,
                    totalPtsTextColor,
                    Utilities.TextAlignment.Center);

                // Points received by current voter
                if (!string.IsNullOrEmpty(entry.Votes[voterIdx]))
                {
                    Utilities.DrawRectangle(
                        canvas,
                        (59f * scaling + xOffset + sizes.FlagOffset + sizes.EntryDetailsWidth,
                         77f * scaling + 35f * scaling * yOffset),
                        24f * scaling,
                        20f * scaling,
                        fillColor: colors.ReceivedPoints);

                    string receivedText = entry.Votes[voterIdx].EndsWith(".0")
                        ? ((int)float.Parse(entry.Votes[voterIdx])).ToString()
                        : entry.Votes[voterIdx];

                    Utilities.DrawText(
                        canvas,
                        (71f * scaling + xOffset + sizes.FlagOffset + sizes.EntryDetailsWidth,
                         87f * scaling + 35f * scaling * yOffset),
                        receivedText,
                        fonts.ReceivedPoints,
                        colors.ReceivedPointsText,
                        Utilities.TextAlignment.Center);
                }

                // Divider line
                Utilities.DrawLine(
                    canvas,
                    (10f * scaling + xOffset, 104.5f * scaling + 35f * scaling * yOffset),
                    (10f * scaling + xOffset + sizes.Rectangle, 104.5f * scaling + 35f * scaling * yOffset),
                    colors.DividerLine,
                    0.5f * scaling);
            }

            // Save image
            string fileName = SanitizeFileName($"{voterIdx + 1} - {config.Contest.VoterNames[voterIdx]}.png");
            string filePath = Path.Combine(config.OutputDir, fileName);

            using var image = SKImage.FromBitmap(bitmap);
            using var data = image.Encode(SKEncodedImageFormat.Png, 100);
            using var stream = File.OpenWrite(filePath);
            data.SaveTo(stream);

            return filePath;
        }

        public static List<string> GenerateContestScoreboards(Config config)
        {
            var fonts = new Fonts(config.FontsConfig);
            var colors = new Colors(config.MainColor, config.AccentColor);

            var results = new string[config.Contest.NumVoters];
            Parallel.For(0, config.Contest.NumVoters, i =>
            {
                results[i] = GenerateScoreboard(config, fonts, colors, i);
            });

            return results.ToList();
        }
    }